### PR TITLE
Implement fn:path#0 and fn:path#1

### PR DESCRIPTION
**Pull request by chris-evolvedbinary**
Implementation of the [`fn:path()`]([https://www.w3.org/TR/xpath-functions-31/#func-path](https://www.w3.org/TR/xpath-functions-31/#func-path)) function.  This implementation requires fixing an [issue](https://github.com/eXist-db/exist/issues/1463) with document nodes.

### Reference:

* https://github.com/eXist-db/exist/issues/3635
* Closes https://github.com/eXist-db/exist/issues/1463

### Type of tests:

* `exist-core/src/test/xquery/xquery3/fnPath.xqm`

--

This open source contribution to the eXist-db project was commissioned by the Office of the Historian, U.S. Department of State, https://history.state.gov/.

### CODE REVIEW

File: exist-core/src/main/java/org/exist/xquery/functions/fn/FunPath.java [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+        if (node.getNodeType() == Type.ATTRIBUTE) {
+            attributeNode = node;
+            node = ((Attr) node).getOwnerElement();
+        } else {
+            attributeNode = null;
+        }
+
+        final StringBuilder value = new StringBuilder();
+        final StringBuilder attributeValue = new StringBuilder();
+        if (node.getNodeType() == Type.TEXT) {
```
Comment (dizzzz) [inline exist-core/src/main/java/org/exist/xquery/functions/fn/FunPath.java:149]: a `switch` would be more correct here?

File: exist-core/src/main/java/org/exist/xquery/functions/fn/FunPath.java [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+            if (nodePosition > 0) {
+                value.append(String.format("/text()[%d]", nodePosition));
+            }
+        } else if (node.getNodeType() == Type.COMMENT) {
+            // For a comment node: comment()[position] where position is an
+            // integer representing the position of the selected node among
+            // its comment node siblings.
+            final int nodePosition = getNodePosition(node);
+            if (nodePosition > 0) {
+                value.append(String.format("/comment()[%d]", nodePosition));
```
Comment (dizzzz) [inline exist-core/src/main/java/org/exist/xquery/functions/fn/FunPath.java:163]: is the 1st position nodePosition 0 or 1 ?
does the XQTS not require `/comment()`  (without indication) to be used iso `/comment()[1]` ?

File: exist-core/src/main/java/org/exist/xquery/functions/fn/FunPath.java [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+            }
+        } else if (node.getNodeType() == Type.PROCESSING_INSTRUCTION) {
+            // For a processing-instruction node: processing-instruction(local)[position]
+            // where local is the name of the processing instruction node and position is
+            // an integer representing the position of the selected node among its
+            // like-named processing-instruction node siblings.
+            final int nodePosition = getNodePosition(node);
+            if (nodePosition > 0) {
+                value.append(String.format("/processing-instruction(%s)[%d]", node.getNodeName(), nodePosition));
```
Comment (dizzzz) [inline exist-core/src/main/java/org/exist/xquery/functions/fn/FunPath.java:172]: same position question

**Comment by dizzzz** [comment]
how to deal with #1463 ? It caused me to hold my PR....

**Review by dizzzz — COMMENTED** [review]
LGTM, some smaller questions

**Comment by adamretter** [comment]
@chris-evolvedbinary Also could you please use the prefixes in your commit messages. If you look at the commit log you will see we use prefixes like [bugfix], [feature], [test], [refactor], etc.

File: exist-core/src/main/java/org/exist/xquery/functions/fn/FunPath.java (hunk 2 above)
Comment (adamretter) [reply exist-core/src/main/java/org/exist/xquery/functions/fn/FunPath.java]: This is passing all XQTS tests for fn-path

**Comment by adamretter** [comment]
> How to deal with https://github.com/eXist-db/exist/issues/1463 ? It caused me to hold my PR....

@dizzzz this will replace that PR and add some additional fixes for fn:root also.

**Pull request merged**